### ISSUE TITLE
PostgreSQLEnumType error "Could not set value of type" with H6 and Panache

### ISSUE THREAD

**Issue by sigi-tw** [vladmihalcea/hypersistence-utils#634]
We are using hypersistence-utils together with Quarkus, Kotlin, Hibernate 6 and Postgresql 15.

After migrating from quarkus 2 and hibernate 5 we now face an enum issue we hadn't had before.

`
Could not set value of type [DemoEnum] : DemoEntity.demoEnum (setter)
`

`
        at org.hibernate.property.access.spi.SetterFieldImpl.set(SetterFieldImpl.java:86)
        at org.hibernate.property.access.spi.EnhancedSetterImpl.set(EnhancedSetterImpl.java:40)
        at org.hibernate.persister.entity.AbstractEntityPersister.setPropertyValues(AbstractEntityPersister.java:4155)
        at org.hibernate.sql.results.graph.entity.AbstractEntityInitializer.initializeEntityInstance(AbstractEntityInitializer.java:844)
        at org.hibernate.sql.results.graph.entity.AbstractEntityInitializer.initializeEntity(AbstractEntityInitializer.java:804)
        at org.hibernate.sql.results.graph.entity.AbstractEntityInitializer.initializeInstance(AbstractEntityInitializer.java:790)
`

We also converted the project into Java and get the same issue.

Java: https://github.com/neukunft/pg-issue-java/tree/main
Kotlin: https://github.com/neukunft/pg-issue

Any suggestions highly appreceated.

**Comment by vladmihalcea** [comment]
From the stack trace, it looks like the problem is in Hibernate core.

Why do you think it's related to Hypersistence Utils?

**Comment by vladmihalcea** [comment]
There's [this test](https://github.com/vladmihalcea/hypersistence-utils/blob/master/hypersistence-utils-hibernate-62/src/test/java/io/hypersistence/utils/hibernate/type/basic/PostgreSQLEnumTest.java) that works fine on Hibernate 6.2.

So, the first step would be to change it to replicate the issue.

Afterward, you need to check what's needed to be changed to make it work.

**Comment by vladmihalcea** [comment]
If it cannot be replicated in Hypersistence Utils, then it might be a Panache or Quarkus issue.

**Comment by vladmihalcea** [comment]
You're welcome.

**Comment by geoand** [comment]
As explained [here](https://github.com/quarkusio/quarkus/issues/33953#issuecomment-1587177469), the problem is actually a `Hypersistence Utils` problem.

**Comment by geoand** [comment]
Does this project have testing infrastructure that allows for separating which classes / jars get loaded by which ClassLoader? If so, I'd be glad to add a test if you point me to the relevant bits and some examples.
If it does not, then you'll forgive for not spending all the time necessary to set that up 😉

**Comment by geoand** [comment]
But apart from that, the analysis of the problem is pretty straightforward, there is nothing really complicated involved (at least from the little time I spent looking into the reproducer) in understanding the general idea of how the problem manifests

**Comment by vladmihalcea** [comment]
> Does this project have testing infrastructure that allows for separating which classes / jars get loaded by which ClassLoader? 

You will find the answer in the `ReflectionUtils` class. 

> But apart from that, the analysis of the problem is pretty straightforward, there is nothing really complicated involved (at least from the little time I spent looking into the reproducer) in understanding the general idea of how the problem manifests

This library is used by millions of developers who use Spring and Spring Boot, and they don't have this issue that you describe. If Quarkus does some atypical things, then we need a replicating test case in Hypersistence Utils to understand what the problem is and how could it be solved without breaking other frameworks using it.

**Comment by geoand** [comment]
> You will find the answer in the ReflectionUtils class.

Well, you are either not giving me all the details or I am not looking in the right place.  

> This library is used by millions of developers who use Spring and Spring Boot, and they don't have this issue that you describe

Of course they don't - those platforms don't support live reloading of application code using a custom ClassLoader. 

> If Quarkus does some atypical things, then we need a replicating test case in Hypersistence Utils to understand what the problem is and how could it be solved without breaking other frameworks that are successfully using Hypersistence Utils

Not objecting in the general sense, but I am pretty that we can all agree that a name does **not** fully identify a class - a class is properly identified by it's name and the ClassLoader that loaded it. Taking this fundamental Java principal in mind, simply using a name as a cache key is not correct in an environment that supports multiple classloaders.

**Comment by vladmihalcea** [comment]
> Not objecting in the general sense, but I am pretty that we can all agree that a name does not fully identify a class - a class is properly identified by it's name and the ClassLoader that loaded it. Taking this fundamental Java principal in mind, simply using a name as a cache key is not correct in an environment that supports multiple classloaders

You can modify that cache to take the `ClassLoader` into consideration as well. I'm sure that this problem you're describing has been already handled in Hibernate or Quarkus, so you can apply the same approach when modifying the [ReflectionUtils](https://github.com/vladmihalcea/hypersistence-utils/blob/b5ba7e7f5af61a9958e0dd677e986c37c85d1d14/hypersistence-utils-hibernate-62/src/main/java/io/hypersistence/utils/hibernate/util/ReflectionUtils.java#L10) class.

**Comment by vladmihalcea** [comment]
There was no question

**Comment by geoand** [comment]
😆 ,

For the record, there certainly was :)

But in any case, I'll take the bait and provide a test sometime this week.

**Comment by vladmihalcea** [comment]
Providing a test case would be the first step to validate the problem.

Afterward, you will also need to provide a fix for this because, other than IBM and Red Hat, it's very unlikely that you will find anyone to provide a fix for this problem in their spare time.

**Comment by geoand** [comment]
FWIW forks, here is the test:

```java
package io.hypersistence.utils.hibernate.util;

import static junit.framework.TestCase.assertEquals;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;
import org.junit.Test;

public class ReflectionUtilsClassLoaderTest {

    @Test
    public void testGetClass() {
        byte[] appClassBytes = getClassBytes(AppClass.class);

        HashMap<String, byte[]> customClasses = new HashMap<>();
        customClasses.put(AppClass.class.getName(), appClassBytes);
        CustomClassLoader cl1 = new CustomClassLoader(this.getClass().getClassLoader(), customClasses);
        CustomClassLoader cl2 = new CustomClassLoader(this.getClass().getClassLoader(), customClasses);

        tryLoadClass(AppClass.class, cl1);
        tryLoadClass(AppClass.class, cl2);

        withNewTCCL(cl1, c -> assertEquals(c.getClassLoader(), cl1));
        withNewTCCL(cl2, c -> assertEquals(c.getClassLoader(), cl2));
    }

    private void withNewTCCL(ClassLoader cl, Consumer<Class<?>> consumer) {
        ClassLoader old =
                Thread.currentThread().getContextClassLoader();

        try {
            Thread.currentThread().setContextClassLoader(cl);
            Class<?> clazz = ReflectionUtils.getClassOrNull(AppClass.class.getName());
            consumer.accept(clazz);
        } finally {
            Thread.currentThread().setContextClassLoader(old); // restore the previous TCCL
        }
    }

    private static void tryLoadClass(Class<AppClass> clazz, CustomClassLoader classLoader) {
        try {
            Class<?> aClass = classLoader.loadClass(clazz.getName());
            assertEquals(aClass.getClassLoader(), classLoader); // ensure that our classloader loaded the class instead of its parent
        } catch (ClassNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    private byte[] getClassBytes(Class<?> clazz) {
        try {
            return this.getClass().getClassLoader().getResourceAsStream(clazz.getName().replace('.', '/') + ".class").readAllBytes();
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    public static class CustomClassLoader extends ClassLoader {

        private final Map<String, byte[]> customClasses;

        private static final AtomicInteger incr = new AtomicInteger(1);
        public CustomClassLoader(ClassLoader parent, Map<String, byte[]> customClasses) {
            super(parent);
            this.customClasses = customClasses;
        }

        @Override
        protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
            Class<?> ex = findLoadedClass(name);
            if (ex != null) {
                return ex;
            }
            if (customClasses.containsKey(name)) {
                return findClass(name);
            }
            return super.loadClass(name, resolve);
        }

        @Override
        protected Class<?> findClass(String name) throws ClassNotFoundException {
            byte[] bytes = customClasses.get(name);
            if (bytes == null) {
                throw new ClassNotFoundException();
            }
            return defineClass(name, bytes, 0, bytes.length);
        }

        @Override
        public String toString() {
            return "CustomClassLoader" + incr.getAndIncrement();
        }
    }

    public static class AppClass {

    }
}

```

**Comment by geoand** [comment]
If you want me to open a PR with this, I certainly can :)

**Comment by geoand** [comment]
🙏

**Comment by vladmihalcea** [comment]
The Class cache is going to stay. However, it will have to be changed just like in [Hazelcast](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/nio/ClassLoaderUtil.java#L54).

And, yes, the Class Cache has a purpose:

https://github.com/hazelcast/hazelcast/issues/1162

**Comment by geoand** [comment]
That's highly debatable. 

Anyway, I demonstrated the problem is clearly with the library (and of course it's a problem that can easily happen to anyone) so up to you folks how and if you want to handle it.

**Comment by vladmihalcea** [comment]
> That's highly debatable (apart for the questionable performance benefit, you're opening yourself up to potential ClassLoader leaks).

As I wrote in [this article](https://vladmihalcea.com/the-performance-penalty-of-class-forname-when-parsing-jpql-and-criteria-queries/), `Class.forName` has been slow for a very long time across multiple Java versions.

This is not a theoretical idea. I had to fix this problem when I was working on Hibernate ORM as it was causing a lot of performance issues:

- [HHH-10746](https://hibernate.atlassian.net/browse/HHH-10746)
- [HHH-1810](https://hibernate.atlassian.net/browse/HHH-1810)
- [HHH-4959](https://hibernate.atlassian.net/browse/HHH-4959)
- [HHH-11246](https://hibernate.atlassian.net/browse/HHH-11246)

Whether this has changed or not in the latest Java versions, and suddenly this problem is gone for good, I don't know. 

But should we take the risk? Of course not!

**Comment by geoand** [comment]
> But should we take the risk? Of course not!

Your project, your rules! 

In any case, the way things are, you **are** risking situations like this one and even worse ones with ClassLoader leaks.

Also FWIW, in Quarkus we have never identitfied `Class.forName` as being a source of performance problems.

**Comment by vladmihalcea** [comment]
On Java 17, `Class.forName` is fast if the classes have already been loaded (notice the difference between p75 and p99):

    2023-06-12 17:46:40,613 INFO  [main]: c.v.b.h.u.ReflectionUtilsTest - type=TIMER, name=transactionTimer, count=1000, min=0.0015999999999999999, max=595.754, mean=0.9082764999999751, stddev=20.024354273700983, median=0.0038, p75=0.0046, p95=0.014799999999999999, p98=1.8236999999999999, p99=2.6454999999999997, p999=595.754, mean_rate=1062.2102555975293, m1=0.0, m5=0.0, m15=0.0, rate_unit=events/second, duration_unit=milliseconds

Since the 60 and 62 modules use Java 17, then the cache could be removed for these modules. The other modules can run on Java 8, which didn't have a very fast `Class.forName` at the time.

**Comment by vladmihalcea** [comment]
Fixed. 

After inspecting the [`ClassUtils`](https://github.com/spring-projects/spring-framework/blob/446b90172bf6b7c2ecdce16f905c5cfdc02335f8/spring-core/src/main/java/org/springframework/util/ClassUtils.java#L109) in Spring, I noticed that they only cache the most common Java classes while leaving the rest uncached. 

Since the ones we load are very specific, the number of classes loaded via Reflection will not be very high, so the cache will probably have very little impact, even on Java 8.
